service-statistics/src/main/java/org/oskari/statistics/user/StatisticalIndicatorServiceMybatisImpl.java

```diff
@@ -94,37 +98,110 @@ public void deleteByUser(long userId) {
+        // Initialize the year dimension as the only one.
         StatisticalIndicatorDataDimension dim = new StatisticalIndicatorDataDimension("year");
-        dim.addAllowedValue(String.valueOf(userIndicator.getYear()));
         ind.getDataModel().addDimension(dim);
         return ind;
     }
+
+    private UserIndicatorDataRow createDBRow(StatisticalIndicator userIndicator) {
+        UserIndicatorDataRow row = new UserIndicatorDataRow();
+        row.title = new JSONObject(userIndicator.getName()).toString();
+        row.description = new JSONObject(userIndicator.getDescription()).toString();
+        row.source = new JSONObject(userIndicator.getSource()).toString();
+        row.published = userIndicator.isPublic();
+        return row;
+    }
+
+    private void addDimension(StatisticalIndicator ind, UserIndicatorDataRow row) {
+        if(ind.getLayer(row.regionsetId) == null) {
+            ind.addLayer(new StatisticalIndicatorLayer(row.regionsetId, ind.getId()));
+        }
+        ind.getDataModel().getDimension("year").addAllowedValue(String.valueOf(row.year));
+    }
+
+    public StatisticalIndicator saveIndicator(StatisticalIndicator ind, long userId) {
+        int id = ConversionHelper.getInt(ind.getId(), -1);
+        UserIndicatorDataRow row = createDBRow(ind);
+        row.userId = userId;
+        row.id = id;
+        if (id != -1) {
+            // check that it's the users indicator
+            StatisticalIndicator existing = findById(id, userId);
+            if (existing == null) {
+                throw new ServiceRuntimeException("Indicator '" + id + "' not found for user: " + userId);
+            }
+        }
+        try (final SqlSession session = factory.openSession()) {
+            if (id != -1) {
+                // update
+                int updated = getMapper(session).updateIndicator(row);
+                if(updated != 1) {
+                    throw new ServiceRuntimeException("Update wasn't successful");
+                }
+            } else {
+                // insert
+                getMapper(session).addIndicator(row);
+            }
+            session.commit();
+        }
+        return findById(id, userId);
+    }
+    public void saveIndicatorData(long indicator, long regionset, int year, String data) {
+        try (final SqlSession session = factory.openSession()) {
+            // remove possible existing data
+            getMapper(session).deleteData(indicator, regionset, year);
+            // insert new data
+            getMapper(session).addData(indicator, regionset, year, data);
+            session.commit();
+        }
+    }
+    public boolean deleteIndicatorData(long indicator, long regionset, int year) {
+        try (final SqlSession session = factory.openSession()) {
+            int updated = getMapper(session).deleteData(indicator, regionset, year);
+            if(updated != 1) {
+                throw new ServiceRuntimeException("Delete wasn't successful");
```

Comment:
I wouldn't expect `public boolean delete()` to throw a RuntimeException in case there was nothing to delete. Instead I'd expect false to be returned, and an Exception to be thrown only in the case there was an actual Exception (like the DB connection unexpectedly error'd or something)